embedded-graphics/src/lib.rs

```diff
@@ -232,18 +237,130 @@ use crate::pixelcolor::PixelColor;
-///     display.draw(egcircle!((32, 32), 10, stroke_color = Some(Gray8::WHITE)));
+///     let mut circle = egcircle!((32, 32), 10, stroke_color = Some(Gray8::WHITE));
+///     circle.draw(&mut display);
 ///
 ///     // Update the display
 ///     display.flush().expect("Failed to send data to display");
 /// }
 /// ```
-pub trait Drawing<C>
+///
+/// ## Hardware Acceleration
+///
+/// In addition to defining `draw_pixel`, an implementation of `DrawTarget` can also provide
+/// alternative implementations of drawing methods for graphical primitives. Here is an example of
+/// how a display with accelerated methods can implement `DrawTarget`:
+///
+/// ```rust
+/// # use embedded_graphics::prelude::*;
+/// # use embedded_graphics::DrawTarget;
+/// # use embedded_graphics::egrectangle;
+/// # use embedded_graphics::primitives::rectangle::Rectangle;
+/// # use embedded_graphics::pixelcolor::{Gray8, GrayColor};
+/// # use embedded_graphics::drawable::Pixel;
+/// #
+/// # struct SPI1;
+/// #
+/// # impl SPI1 {
+/// #     pub fn send_bytes(&self, buf: &[u8]) -> Result<(), ()> {
+/// #         Ok(())
+/// #     }
+/// # }
+/// #
+/// /// A fake display 64px x 64px where each pixel is stored as a single `u8`
+/// struct FastExampleDisplay {
+///     framebuffer: [u8; 64 * 64],
+///     iface: SPI1,
+/// }
+///
+/// impl FastExampleDisplay {
+///     /// Send buffer to the display
+///     pub fn flush(&self) -> Result<(), ()> {
+///         self.iface.send_bytes(&self.framebuffer)
+///     }
+///
+///     /// A HW-accelerated method for drawing rectangles
+///     pub fn fast_rectangle(&self, rect: &Rectangle<Gray8>) {
+///         // Does some speedy drawing
+///     }
+/// }
+///
+/// impl DrawTarget<Gray8> for FastExampleDisplay {
+///     /// Draw a `pixel` that has a colour defined as `Gray8`
+///     fn draw_pixel(&mut self, pixel: Pixel<Gray8>) {
+///         let Pixel(coord, color) = pixel;
+///         // Place an (x, y) pixel at the right index in the framebuffer
+///         let index = coord[0] + (coord[1] * 64);
+///         self.framebuffer[index as usize] = color.luma();
+///     }
+///
+///     /// Use the accelerated method when drawing rectangles
+///     fn draw_rectangle(&mut self, item: &Rectangle<Gray8>) {
+///         self.fast_rectangle(item);
+///     }
+/// }
+///
+/// fn main() {
+///     let mut display = FastExampleDisplay {
+///         framebuffer: [0; 4096],
+///         iface: SPI1
+///     };
+///
+///     // Draw a rectangle from (10, 20) to (30, 40) with a white stroke
+///     let mut rect = egrectangle!((10, 20), (30, 40), stroke_color = Some(Gray8::WHITE));
+///     rect.draw(&mut display); // Uses the accelerated draw_rectangle function
+///
+///     // Update the display
+///     display.flush().expect("Failed to send data to display");
+/// }
+///
+/// ```
+pub trait DrawTarget<C>
 where
     C: PixelColor,
 {
+    /// Draw a pixel on the display
```

Comment:
Should there be an extra note here explaining that some displays will require a `flush()` of some sort to actually write the framebuffer? Others will work in immediate mode so won't require the `flush()` call. Should this be commented on here, or left to display drivers to describe?

Reply:
Hmm, that's a good point. That seems like a reasonable thing to caution about in the documentation-- it's good for consumers of an eg-compatible library to know about as well.

In that vein, should there be a `flush()` method on `DrawTarget`? It seems like it should be required, and immediate mode drivers can just implement it as a noop function.

Reply:
> In that vein, should there be a flush() method on DrawTarget? It seems like it should be required, and immediate mode drivers can just implement it as a noop function.

I don't think so, that seems pretty driver dependent to me.

Reply:
@jamwaffles I just pushed a commit that added a default flush method to a `DrawTarget` that is just a noop. I can revert this if you would like; however, I thought it might make sense to include flushing as part of the interface since it would not be possible to write portable code for flushing displays if that functionality were not exposed. Furthermore, this seems to shadow what is done in the stdlib `io` crate, which requires a flushing method even though not all `Write` objects necessarily need it.

Reply:
Ok, that's fair. I didn't know about `flush` in `io` but that does make sense, particularly with a default impl.